### PR TITLE
Redo request decoding

### PR DESCRIPTION
**Pull request by jinty**
It looks to me that the original attempt at porting zope.publisher to python3 didn't take into account that FieldStorage started decoding the input in python 3:

https://hg.python.org/cpython/rev/d864328a74e4

In fact the existing decoding code doesn't work at all if you go outside latin-1.

This leaves zope.publisher in a bad place trying to re-do what cgi.py had already done when running in Python 3. I've attempted to kludge around that, trying to maintain backwards compat as well as not redo the decoding cgi.FieldStorage has already done.

This improves things, but is by no means perfect. It's good enough for me and I don't really have time to resolve this fully. A good start for future work would be more tests of the behaviour of IUserPreferredCharsets under non utf-8 environments.

### CODE REVIEW

**Comment by icemac** [comment]
@jinty I stumbled over your PR from 2016. Is it still valid? Should we get it merged?

**Comment by jinty** [comment]
On Thu, Sep 28, 2017 at 05:54:26AM -0700, Michael Howitz wrote:
> @jinty I stumbled over your PR from 2016. Is it still valid? Should we get it merged?

Well, we use it in production every day as a stopgap to better things
while we phase out zope.publisher :)

I don't see how anyone could use zope.publisher under python 3 in a
non-ascii environment without something like it.

However my patch is ugly and not really the right way to solve the
problem. I would be content with just leaving it as a pull request for
anyone who needs it to find/use it.


> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/zopefoundation/zope.publisher/pull/15#issuecomment-332827197

-- 
Brian Sutherland

**Comment by janwijbrand** [comment]
Just a quick comment (I hope to get back later with an more informed opinion/suggestion): I'm with @jinty on this one: `zope.publisher` essentially does not work on Python 3. I'll now fork it and see if, at least in our stack, I can get it to work again.

**Comment by mgedmin** [comment]
OMG when I filed #40 and #41 I was wondering "am I crazy? zope.publisher is totally broken on Python 3!  how is everyone else using it?", and I _completely forgot_ about this PR!

File: src/zope/publisher/browser.py [hunk 1]
```diff
@@ -247,16 +248,17 @@ def _createResponse(self):
...
             envadapter = IUserPreferredCharsets(self)
             self.charsets = envadapter.getPreferredCharsets() or ['utf-8']
             self.charsets = [c for c in self.charsets if c != '*']
+        if not PYTHON2 and self.charsets and self.charsets[0] == 'utf-8':
+            # optimization: we are trying to decode something cgi.FieldStorage already
+            # decoded for us, let's just return it rather than waste time decoding
+            return text
+        if not PYTHON2:
+            # undo what cgi.FieldStorage did and maintain backwards compat
+            text = text.encode('utf-8')
```
Comment (mgedmin) [inline src/zope/publisher/browser.py]: This was a bit hard to me to follow, with the partially overlapping if conditions.  I would suggest

```python
if not PYTHON2:
    if self.charsets and self.charsets[0] == 'utf-8':
        # optimization: ...
        return text
    # undo what cgi.FieldStorage did....
    text = text.encode('utf-8')
```

File: src/zope/publisher/tests/test_browserrequest.py [hunk 2]
```diff
@@ -236,6 +237,35 @@ def testLargePostValue(self):
...
         request.processInputs()
 
+    def test_multipart_form_decoding_in_UTF8(self):
+        street = u'NOT latin-1: 汉语/漢語'
+        body = u"""---123
+Content-Disposition: form-data; name="street"
+
+%s
+---123--
+""" % street
```
Comment (mgedmin) [inline src/zope/publisher/tests/test_browserrequest.py]: A textwrap.dedent() would do wonders here.

File: src/zope/publisher/tests/test_browserrequest.py [hunk 3]
```diff
@@ -283,18 +313,23 @@ def testForm(self):
...
-            'QUERY_STRING': 'a=5&b:int=6&street=' + encoded
+            'QUERY_STRING': qs
             }
         request = self._createRequest(extra)
         # many mainstream browsers do not send HTTP_ACCEPT_CHARSET
         del request._environ['HTTP_ACCEPT_CHARSET']
         publish(request)
         self.assertTrue(isinstance(request.form[_u("street")], unicode))
-        self.assertEqual(_u("K\xf6hlerstra\xdfe"), request.form['street'])
+        self.assertEqual(street, request.form['street'])
```
Comment (mgedmin) [inline src/zope/publisher/tests/test_browserrequest.py]: Are changes to this test really necessary?  The original `encoded` was already UTF-8-as-bytes on PYTHON2 and UTF-8-mangled-into-Latin-1 on not PYTHON2.

File: src/zope/publisher/browser.py (hunk 1 above)
Comment (mgedmin) [reply src/zope/publisher/browser.py]: Also, I'm guessing this is lossy?  `cgi.FieldStorage()` defaults to `encoding='UTF-8', errors='replace'` and once you've replaced invalid UTF-8 characters, there's no way of getting them back.

We could pass `latin-1` to ZopeFieldStorage() and then do our own charset conversion here.

Comment (mgedmin) [reply src/zope/publisher/browser.py]: Ah, I've also discovered cases where cgi.FieldStorage gives us `b''` as the value, when the POSTed form contains a blank file field (in a request generated by zope.testbrowser/WebTest).

**Comment by mgedmin** [comment]
I've dusted this PR off and prepared a new version with some additional fixes in https://github.com/zopefoundation/zope.publisher/pull/46.

**Pull request closed without merging**